[staging] Build failure: protobuf_25

### Nixpkgs version

- Unstable (26.05)

### Steps to reproduce

nix build nixpkgs/staging#protobuf_25

### Can Hydra reproduce this build failure?

Hydra doesn’t try to build the package.

### Link to Hydra build job

_No response_

### Relevant log output

```console
-- Build files have been written to: /build/source/build
cmake: enabled parallel building
cmake: enabled parallel installing
Running phase: buildPhase
@nix { "action": "setPhase", "phase": "buildPhase" }
build flags: -j10 SHELL=/nix/store/w3c7v5vxfvyr9kk8mdbdp2jng1ffrd01-bash-5.3p9/bin/bash
[  0%] Building CXX object third_party/utf8_range/CMakeFiles/utf8_validity.dir/utf8_validity.cc.o
[  0%] Building C object third_party/utf8_range/CMakeFiles/utf8_range.dir/naive.c.o
[  0%] Building C object third_party/utf8_range/CMakeFiles/utf8_range.dir/range2-sse.c.o
[  1%] Building C object third_party/utf8_range/CMakeFiles/utf8_range.dir/range2-neon.c.o
/build/source/third_party/utf8_range/utf8_validity.cc: In function 'uint64_t utf8_range::{anonymous}::UNALIGNED_LOAD64(const void*)':
/build/source/third_party/utf8_range/utf8_validity.cc:37:3: error: 'memcpy' was not declared in this scope
   37 |   memcpy(&t, p, sizeof t);
      |   ^~~~~~
/build/source/third_party/utf8_range/utf8_validity.cc:24:1: note: 'memcpy' is defined in header '<cstring>'; this is probably fixable by adding '#include <cstring>'
   23 | #include "absl/strings/ascii.h"
  +++ |+#include <cstring>
   24 | #include "absl/strings/string_view.h"
[  1%] Linking C static library libutf8_range.a
make[2]: *** [third_party/utf8_range/CMakeFiles/utf8_validity.dir/build.make:79: third_party/utf8_range/CMakeFiles/utf8_validity.dir/utf8_validity.cc.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:646: third_party/utf8_range/CMakeFiles/utf8_validity.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[  1%] Built target utf8_range
make: *** [Makefile:146: all] Error 2
```

### Additional context

_No response_

### System metadata


 - system: `"aarch64-linux"`
 - host os: `Linux 6.19.5, NixOS, 26.05 (Yarara), 26.05.20260227.f945716`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Lix, like Nix) 2.95.0-pre20260226-dev_bcab3d5
System type: aarch64-linux
Additional system types: 
Features: gc, signed-caches
System configuration file: /etc/nix/nix.conf
User configuration files: /home/yuka/.config/nix/nix.conf:/etc/xdg/nix/nix.conf:/home/yuka/.nix-profile/etc/xdg/nix/nix.conf:/nix/profile/etc/xdg/nix/nix.conf:/home/yuka/.local/state/nix/profile/etc/xdg/nix/nix.conf:/etc/profiles/per-user/yuka/etc/xdg/nix/nix.conf:/nix/var/nix/profiles/default/etc/xdg/nix/nix.conf:/run/current-system/sw/etc/xdg/nix/nix.conf
Store directory: /nix/store
State directory: /nix/var/nix
Data directory: /nix/store/nwj1bjc6nx4swi5ha1bchy7s43w3xi57-lix-2.95.0-pre20260226-dev_bcab3d5/share`
 - nixpkgs: `/nix/store/p8ajy5jkx5sdls1b5cbcsh7k4ixc3m2w-source`

### Notify maintainers

@GaetanLepage 

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+label%3A%220.kind%3A+build+failure%22).
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/development/libraries/protobuf/generic.nix
+++ b/pkgs/development/libraries/protobuf/generic.nix
@@ -85,6 +85,10 @@ stdenv.mkDerivation (finalAttrs: {
         --replace-fail \
           '#define UPB_LINKARR_SENTINEL UPB_RETAIN __attribute__((weak, used))' \
           '#define UPB_LINKARR_SENTINEL            __attribute__((weak, used))'
+    ''
+    # Fix gcc15 build failures due to missing <cstring>
+    + lib.optionalString ((lib.versions.major version) == "25") ''
+      sed -i '1i #include <cstring>' third_party/utf8_range/utf8_validity.cc
     '';
 
   preHook = ''

--- a/pkgs/development/python-modules/protobuf/4.nix
+++ b/pkgs/development/python-modules/protobuf/4.nix
@@ -51,23 +51,24 @@ buildPythonPackage {
     fi
   '';
 
-  # Remove the line in setup.py that forces compiling with C++14. Upstream's
-  # CMake build has been updated to support compiling with other versions of
-  # C++, but the Python build has not. Without this, we observe compile-time
-  # errors using GCC.
-  #
-  # Fedora appears to do the same, per this comment:
-  #
-  #   https://github.com/protocolbuffers/protobuf/issues/12104#issuecomment-1542543967
-  #
-  postPatch = ''
-    sed -i "/extra_compile_args.append('-std=c++14')/d" setup.py
-
+  postPatch =
+    # Remove the line in setup.py that forces compiling with C++14. Upstream's
+    # CMake build has been updated to support compiling with other versions of
+    # C++, but the Python build has not. Without this, we observe compile-time
+    # errors using GCC.
+    #
+    # Fedora appears to do the same, per this comment:
+    #
+    #   https://github.com/protocolbuffers/protobuf/issues/12104#issuecomment-1542543967
+    ''
+      sed -i "/extra_compile_args.append('-std=c++14')/d" setup.py
+    ''
     # The former function has been renamed into the latter in Python 3.12.
     # Does not apply to all protobuf versions, hence --replace-warn.
-    substituteInPlace google/protobuf/internal/json_format_test.py \
-      --replace-warn assertRaisesRegexp assertRaisesRegex
-  '';
+    + ''
+      substituteInPlace google/protobuf/internal/json_format_test.py \
+        --replace-warn assertRaisesRegexp assertRaisesRegex
+    '';
 
   nativeBuildInputs = lib.optional isPyPy tzdata;
 
@@ -85,21 +86,29 @@ buildPythonPackage {
   ]
   ++ lib.optionals (lib.versionAtLeast protobuf.version "22") [ numpy ];
 
-  disabledTests =
-    lib.optionals isPyPy [
-      # error message differs
-      "testInvalidTimestamp"
-      # requires tracemalloc which pypy does not implement
-      # https://foss.heptapod.net/pypy/pypy/-/issues/3048
-      "testUnknownFieldsNoMemoryLeak"
-      # assertion is not raised for some reason
-      "testStrictUtf8Check"
-    ]
-    ++ lib.optionals stdenv.hostPlatform.is32bit [
-      # OverflowError: timestamp out of range for platform time_t
-      "testTimezoneAwareDatetimeConversionWhereTimestampLosesPrecision"
-      "testTimezoneNaiveDatetimeConversionWhereTimestampLosesPrecision"
-    ];
+  disabledTests = [
+    # TypeError: np.False_ has type <class 'numpy.bool'>,
+    # but expected one of: (<class 'bool'>, <class 'int'>)
+    "NumpyBoolProtoTest"
+  ]
+  ++ lib.optionals isPyPy [
+    # error message differs
+    "testInvalidTimestamp"
+    # requires tracemalloc which pypy does not implement
+    # https://foss.heptapod.net/pypy/pypy/-/issues/3048
+    "testUnknownFieldsNoMemoryLeak"
+    # assertion is not raised for some reason
+    "testStrictUtf8Check"
+  ]
+  ++ lib.optionals stdenv.hostPlatform.is32bit [
+    # OverflowError: timestamp out of range for platform time_t
+    "testTimezoneAwareDatetimeConversionWhereTimestampLosesPrecision"
+    "testTimezoneNaiveDatetimeConversionWhereTimestampLosesPrecision"
+  ]
+  ++ lib.optionals (pythonAtLeast "3.14") [
+    # AssertionError: "year (0 )?is out of range" does not match "year must be in 1..9999, not 0"
+    "testInvalidTimestamp"
+  ];
 
   disabledTestPaths =
     lib.optionals (lib.versionAtLeast protobuf.version "23") [
@@ -115,6 +124,9 @@ buildPythonPackage {
     ]
     ++ lib.optionals (lib.versionAtLeast protobuf.version "25") [
       "minimal_test.py" # ModuleNotFoundError: No module named 'google3'
+
+      # ImportError: cannot import name 'self_recursive_pb2' from 'google.protobuf.internal'
+      "google/protobuf/internal/message_test.py"
     ];
 
   pythonImportsCheck = [

